plugins/c8y_configuration_plugin/src/main.rs

```diff
@@ -0,0 +1,185 @@
+`c8y_configuration_plugin` subscribes then to `c8y/s/ds` listening for configuration operation requests (messages `524` and `526`).
+notifying the Cumulocity tenant of their progress (messages `501`, `502` and `503`).
+
+The thin-edge `CONFIG_DIR` is used to find where:
+  * to store temporary files on download: `tedge config get tmp.path`,
+  * to log operation errors and progress: `tedge config get log.path`,
+  * to connect the MQTT bus: `tedge config get mqtt.port`."#;
+
+#[derive(Debug, clap::Parser)]
+#[clap(
+name = clap::crate_name!(),
+version = clap::crate_version!(),
+about = clap::crate_description!(),
+after_help = AFTER_HELP_TEXT
+)]
+pub struct ConfigPluginOpt {
+    /// Turn-on the debug log level.
+    ///
+    /// If off only reports ERROR, WARN, and INFO
+    /// If on also reports DEBUG and TRACE
+    #[clap(long)]
+    pub debug: bool,
+
+    /// Create supported operation files
+    #[clap(short, long)]
+    pub init: bool,
+
+    #[clap(long = "config-dir", default_value = DEFAULT_TEDGE_CONFIG_PATH)]
+    pub config_dir: PathBuf,
+
+    #[clap(long = "config-file", default_value = DEFAULT_PLUGIN_CONFIG_FILE_PATH)]
+    pub config_file: PathBuf,
+}
+
+async fn create_mqtt_client(
+    tedge_config: &TEdgeConfig,
+) -> Result<mqtt_channel::Connection, anyhow::Error> {
+    let mqtt_port = tedge_config.query(MqttPortSetting)?.into();
+    let mqtt_config = mqtt_channel::Config::default()
+        .with_port(mqtt_port)
+        .with_subscriptions(mqtt_channel::TopicFilter::new_unchecked(
+            C8yTopic::SmartRestRequest.as_str(),
+        ));
+
+    let mqtt_client = mqtt_channel::Connection::new(&mqtt_config).await?;
+    Ok(mqtt_client)
+}
+
+pub async fn create_http_client(
+    tedge_config: &TEdgeConfig,
+) -> Result<JwtAuthHttpProxy, anyhow::Error> {
+    let mut http_proxy = JwtAuthHttpProxy::try_new(tedge_config).await?;
+    let () = http_proxy.init().await?;
+    Ok(http_proxy)
+}
+
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+    let config_plugin_opt = ConfigPluginOpt::parse();
+    tedge_utils::logging::initialise_tracing_subscriber(config_plugin_opt.debug);
+
+    if config_plugin_opt.init {
+        init(config_plugin_opt.config_dir)?;
+        return Ok(());
+    }
+
+    // Load tedge config from the provided location
+    let tedge_config_location =
+        tedge_config::TEdgeConfigLocation::from_custom_root(config_plugin_opt.config_dir);
+    let config_repository = tedge_config::TEdgeConfigRepository::new(tedge_config_location.clone());
+    let tedge_config = config_repository.load()?;
+
+    // Create required clients
+    let mut mqtt_client = create_mqtt_client(&tedge_config).await?;
+    let mut http_client = create_http_client(&tedge_config).await?;
+
+    let plugin_config = PluginConfig::new(config_plugin_opt.config_file);
+
+    // Publish supported configuration types
+    let msg = plugin_config.to_message()?;
```

Comment:
Do we resend this message when the plugin configuration is changed?

Reply:
We should - however not yet using inotify.